pkg/reconciler/workload/resource/resource_controller.go

```diff
@@ -98,19 +98,21 @@ func NewController(
 
 		getSyncTargetFromKey: func(syncTargetKey string) (*workloadv1alpha1.SyncTarget, bool, error) {
 			syncTargets, err := indexers.ByIndex[*workloadv1alpha1.SyncTarget](syncTargetInformer.Informer().GetIndexer(), bySyncTargetKey, syncTargetKey)
-			if err != nil && !errors.IsNotFound(err) {
-				return nil, false, err
+			if errors.IsNotFound(err) || len(syncTargets) == 0 {
+				// check cache before deciding what to return
+				syncTargets, err = indexers.ByIndex[*workloadv1alpha1.SyncTarget](globalSyncTargetInformer.Informer().GetIndexer(), bySyncTargetKey, syncTargetKey)
+				if errors.IsNotFound(err) || len(syncTargets) == 0 {
+					// still no SyncTarget, so return empty results.
+					return nil, false, nil
```

Comment:
If I'm not mistaken `err` can be non-nil here ( if `len(syncTargets) == 0` ), but a `nil` error will be returned.

Reply:
You're right, the error value would be lost in this code.

I think `if errors.IsNotFound(err) || (len(syncTargets) == 0 && err == nil)` could cover the case in which `len(syncTargets) == 0` and some other error was returned.

I was trying to reduce the number of if statements involved in returning values, but I can split them again if that's easier to follow.

Reply:
My main point it to avoid loosing the error in some cases

Reply:
The error here would be a missing index, which would be a programmer error & not equivalent to the case where no `SyncTarget`s exist